p2p/connectionpool/connectionpool.go

```diff
@@ -84,30 +80,23 @@ func (cp *ConnectionPool) OnClosedConnection(ctx context.Context, cwe net.Connec
 }
 
 func (cp *ConnectionPool) isShuttingDown() bool {
-	var isd bool
-	cp.connMutex.RLock()
-	isd = cp.shutdown
-	cp.connMutex.RUnlock()
-	return isd
+	select {
+	case <-cp.shutdownCtx.Done():
+		return true
+	default:
+	}
+	return false
 }
 
 // Shutdown gracefully shuts down the ConnectionPool:
 // - Closes all open connections
 // - Waits for all Dial routines to complete and unblock any routines waiting for GetConnection
 func (cp *ConnectionPool) Shutdown() {
-	cp.shutdownCancel()
-	cp.connMutex.Lock()
-	if cp.shutdown {
-		cp.connMutex.Unlock()
-		cp.logger.Error("shutdown was already called")
-		return
-	}
-	cp.shutdown = true
-	cp.connMutex.Unlock()
-
-	cp.dialWait.Wait()
-	// we won't handle the closing connection events for these connections since we exit the loop once the teardown is done
-	cp.closeConnections()
+	cp.shutdownOnce.Do(func() {
```

Comment:
I guess this is being called on `switch.go`.. I thought the advantage of using a context means we don't have to call this explicitly  but that will probably require a running routine or maybe calling that once when isShuttingDown goes down the `true` path, consider switching to one of those.. context is passed down so we could cancel it only once above and it will trigger shutting down all the way down every component.

Reply:
that's the route (i hope) we will eventually take. however, at this time, it seems `Switch.Shutdown()` dictates the sequence of shutting down for p2p components. i don't want to make too much change in one PR. will experiment with the paradigm you described in a later PR.